### PR TITLE
Hotfix to add releaseTime to adpupa converter

### LINKED ISSUE #1322: Add releaseTime to NCEP ADPUPA converter
## Description
The variable releaseTime is required by Skylab, however, IODA files produced by the NCEP ADPUPA converter do not contain this information. The variable should be added to the converter output

## Requirements

Determine release time from the prepbufr data

## Acceptance Criteria (Definition of Done)
code outputs ioda file with the release time included and the file is accepted by Skylab

## Dependencies
none

### PR DESCRIPTION
**Pull request by haydenlj**
## Description

Temporary hotfix to add releaseTime to files output by the new adpupa prepbufr to ioda converter so that output files can be used by Skylab. Can be reverted once the planned more robust additions have been made

## Issue(s) addressed

Resolves #1322 

## Dependencies

none
## Impact

Output IODA files can be used in Skylab

## Checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have run the unit tests before creating the PR

### CODE REVIEW

**Review by rmclaren — CHANGES_REQUESTED** [review]
I don't believe this is what you want... The reason is that now you've mixed your data set with fields that are grouped by prepbufrDataLevelCategory (latitude, longitude, stationElevation... and so on) with things that are not grouped (timeOffset, specificHumidity)... The problem here is that you've lost your ability to correlate between these fields (for example you can no longer tell what latitude correlates with specific humidity). You have to be consistent with grouping to make datasets that make sense.

**Comment by haydenlj** [comment]
> I don't believe this is what you want... The reason is that now you've mixed your data set with fields that are grouped by prepbufrDataLevelCategory (latitude, longitude, stationElevation... and so on) with things that are not grouped (timeOffset, specificHumidity)... The problem here is that you've lost your ability to correlate between these fields (for example you can no longer tell how latitude, longitude lines up with specific humidity). You have to be consistent with grouping to make datasets that make sense.

@rmclaren Ah ok, in that case I misunderstand what the bufr `get` function is doing. So it is sorting by prepbufrDataLevelCategory rather than just flattening along that axis?

File: test/testinput/bufr_ncep_prepbufr_adpupa.py
Comment (PraveenKumar-NOAA) [inline test/testinput/bufr_ncep_prepbufr_adpupa.py:67]: You have to be consistent with grouping as said by @rmclaren.

File: test/testinput/bufr_ncep_prepbufr_adpupa.py
Comment (PraveenKumar-NOAA) [inline test/testinput/bufr_ncep_prepbufr_adpupa.py:84]: You have to be consistent with grouping as said by @rmclaren.

**Review by PraveenKumar-NOAA — COMMENTED** [review]
Only the grouping issues, rest looks fine.

**Review by huishao-r — COMMENTED** [review]
I can't comment on other parts regarding the grouping. But the release time seems fine with me.

File: test/testinput/bufr_ncep_prepbufr_adpupa.py
Comment (haydenlj) [inline test/testinput/bufr_ncep_prepbufr_adpupa.py:72]: @emilyhcliu @PraveenKumar-NOAA @rmclaren Could you clarify what this variable is? When using a 00Z prepbufr file, the times in this variable end up being ~21Z

Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_prepbufr_adpupa.py:72]: @haydenlj is profile level cycle-time, which is used for dateTime calculation. 

| HRDR     | 004218 | PROFILE LVL TIME-CYCLE (FOR RAOB/PIBAL, BASED ON B DFT)  |

You can find more details in the following table: 
https://github.com/JCSDA-internal/ioda-converters/blob/develop/doc/bufr_table_samples/adpupa.prepbufr.table.txt

Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_prepbufr_adpupa.py:72]: @haydenlj @emilyhcliu for 00Z prepbufr file, the cycle starts at 21Z and end ends at 03Z.

**Comment by BenjaminRuston** [comment]
let's not merge this until we have NOAA blessing that it's working for them as well

**Comment by gthompsnJCSDA** [comment]
@haydenlj  Could you also try running the ioda-validate.x program on the output file that gets created?  That will help ensure we also don't have something like `MetaData/height` as an integer when it should be a float.  Or plenty of other possible examples as well.  This just seems like the right time to catch anything we can that could become technical debt later.

**Comment by BenjaminRuston** [comment]
we do not think there is any grouping issue and we would like to have NOAA EMC test this PR to see if it still reproduces the answer you expect.

@emilyhcliu please advise

**Comment by PraveenKumar-NOAA** [comment]
> we do not think there is any grouping issue and we would like to have NOAA EMC test this PR to see if it still reproduces the answer you expect.
> 
> @emilyhcliu please advise

@BenjaminRuston I will be happy to test this out.

**Comment by PraveenKumar-NOAA** [comment]
> we do not think there is any grouping issue and we would like to have NOAA EMC test this PR to see if it still reproduces the answer you expect.
> 
> @emilyhcliu please advise

@BenjaminRuston I tested this PR/ Python script and it is reproducing the expected results. If @emilyhcliu @rmclaren are happy with this new format, I will approve this PR.

**Comment by PraveenKumar-NOAA** [comment]
> @haydenlj Could you also try running the ioda-validate.x program on the output file that gets created? That will help ensure we also don't have something like `MetaData/height` as an integer when it should be a float. Or plenty of other possible examples as well. This just seems like the right time to catch anything we can that could become technical debt later.

@gthompsnJCSDA I ran ioda-validate.x program on the output file and there were a few errors/ warnings in the final results. A few new prepBUFR variables like prepbufrDataLevelCategory (CAT) and temperatureEventCode (TPC) need to be added to the ObsSpace.yaml.  I will create an issue for this.

**Comment by gthompsnJCSDA** [comment]
> > @haydenlj Could you also try running the ioda-validate.x program on the output file that gets created? That will help ensure we also don't have something like `MetaData/height` as an integer when it should be a float. Or plenty of other possible examples as well. This just seems like the right time to catch anything we can that could become technical debt later.
> 
> @gthompsnJCSDA I ran ioda-validate.x program on the output file and there were a few errors/ warnings in the final results. A few new prepBUFR variables like prepbufrDataLevelCategory (CAT) and temperatureEventCode (TPC) also need to be added to the ObsSpace.yaml. I will create an issue for this.

It seems we must already have something similar to DataLevelCategory - could you try to re-use something that is similar, such as `verticalSignificance`?  And, isn't `temperatureEvent` already in the ObsSpace.yaml and was already exactly designed for radiosonde data I thought (similar to windEvent and moistureEvent, etc.).

**Comment by BenjaminRuston** [comment]
@emilyhcliu looks like @PraveenKumar-NOAA has tested and happy with result are you ready to move forward

**Comment by PraveenKumar-NOAA** [comment]
> 

@gthompsnJCSDA thanks, there is no need to add any extra variables to the ObsSpace.yaml for the adpupa prepbufr. 

@haydenlj please use ObsValue/_verticalSignificance_ for MetaData/prepbufrDataLevelCategory (CAT) and QCFlags/_qualityFlags_ for the MetaData/temperatureEventCode (TPC), then there will be no error in your ioda-validate.x run, you just need to take care of a few warnings. I can do that easily.

**Comment by haydenlj** [comment]
> QCFlags/_qualityFlags_

Ok thanks @PraveenKumar-NOAA I updated the names

**Comment by BenjaminRuston** [comment]
@haydenlj looks like the testoutput may need to be updated:
```
Write data to variables
DIFFER : NUMBER OF GROUPS : 4 ("/") <> 3 ("/")
DIFFER : VARIABLE : prepbufrDataLevelCategory : DOES NOT EXIST IN "testrun/prepbufr_adpupa_api.nc"
DIFFER : VARIABLE : temperatureEventCode : DOES NOT EXIST IN "testrun/prepbufr_adpupa_api.nc"
DIFFER : VARIABLE : verticalSignificance : DOES NOT EXIST IN "testoutput/prepbufr_adpupa_api.nc"
DIFFER : Failed to find group /QCFlags in file "testoutput/prepbufr_adpupa_api.nc"
```

**Comment by emilyhcliu** [comment]
@haydenlj Could you give me the definition of the releaseTime?  Is it the earliest time from dateTime of each profile?

**Review by PraveenKumar-NOAA — APPROVED** [review]
Looks good to me.

File: test/testinput/bufr_ncep_prepbufr_adpupa.py
Comment (BenjaminRuston) [inline test/testinput/bufr_ncep_prepbufr_adpupa.py:76]: @emilyhcliu here is the definition of the variable that is used to define the release time which is the first entry of the times array

**Review by BenjaminRuston — APPROVED** [review]
@haydenlj had to make one more change to make sure it ran through Skylab

**Pull request merged**
